File: 2024/day/2/main.py
from itertools import pairwise


with open("input") as f:
    reports = [[int(n) for n in ln.strip().split()] for ln in f]


def safe1(levels: list[int]) -> bool:
    diffs = [a - b for a, b in pairwise(levels)]
    increasing = all(-3 <= d <= -1 for d in diffs)
    decreasing = all(1 <= d <= 3 for d in diffs)
    return increasing or decreasing


def safe2(levels: list[int]) -> bool:
    return any(safe1(levels[:i] + levels[i + 1 :]) for i in range(len(levels)))


ans1 = sum(safe1(report) for report in reports)
print(ans1)

ans2 = sum(safe2(report) for report in reports)
print(ans2)

File: 2024/day/2/Main.hs
safe1 :: [Int] -> Bool
safe1 levels = increasing || decreasing
  where
    diffs = zipWith (-) levels (drop 1 levels)
    increasing = all (\d -> 1 <= d && d <= 3) diffs 
    decreasing = all (\d -> -3 <= d && d <= -1) diffs
    
safe2 :: [Int] -> Bool
safe2 levels = any (\i -> safe1 (take i levels ++ drop (i + 1) levels)) [0..length levels - 1]

main = do
    input <- readFile "input"
    let reports = map (map read . words) (lines input)
    
    -- Part 1
    print $ length [report | report <- reports, safe1 report]
    
    -- Part 2
    print $ length [report | report <- reports, safe2 report]